### ISSUE TITLE
Is it a necessity for Dokan to call the same callback 2, 3, 4 + times in a row?

### ISSUE THREAD

**Issue by kdschlosser** [dokan-dev/dokany#1153]
# Environment
 * Windows version: 11
 * Processor architecture: x64
 * Dokany version: 2.0.6.1000
 * Library type (Dokany/FUSE): Dokany
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description below.
  - [x] I have updated Dokany to the [latest version](https://github.com/dokan-dev/dokany/releases) and have reboot my computer after.
  - [x] I tested one of the last [snapshot](https://github.com/dokan-dev/dokany/wiki/Build#user-snapshot) from appveyor CI

# Description

I have picked up working on the python binding for Dokany. It has been a number of years since I had worked on it. I have gotten the binding to partially work and one of the things that is causing me a lot of grief is quite often the same callback gets multiple times requesting the same information. This happens in succession, one right after the other with no other callback getting called. The same information is being requested each time. This would cause serious problems if a slower device is being communicated with. Something like an MCU over serial. It also has a side impact of making it horrible to try and manage if Dokany is not set to use a single thread or if mutexes are not used to stop access to the same callback concurrently.

# Logs
Here is a log from a ported version of the mirror example. The log formatting has been changed to make it easier for me to understand what is happening. There is also a shared "lock" across all of the callbacks allowing only one callback to run at a time. which keeps the logging information aligned properly.

You can see in the logs some repetition where the same callback is called multiple times in a row but you can also see where there are groups of callbacks that repeat. That log is from expanding the drive and then clicking on a single folder. The program bombs because of it looking for desktop.ini which is not there and I don't know why it is even looking for it. I never get a list of files in any of the sub folders only the folders and files in the root of the drive get displayed properly.

I can also give the ported mirror example code as it is really not that far off from what the original is. the "core" portion of it has remained the same. There is really only the locks and the changing to debugging output that has been altered from the original other than changing the syntax. The naming convention has not changed much if any.

src_path = the "local" path or real path to the file/folder
dst_path = relative path to the mounted drive

[scratch_529.txt](https://github.com/dokan-dev/dokany/files/11717183/scratch_529.txt)

**Comment by Liryna** [comment]
Hi @kdschlosser ,

I believe what you are seeing is just explorer trying to see if the file exists which is a normal behavior.
Dokan just forward the requests sent by softwares and system.
There might be cases when explorer really spam abnormally but this would only happen if the response to those requests is invalid somehow.

Regarding slow access like serial port, it is the filesystem implementation that should cache the information. Dokan could do the caching but that's more complicated (there is a ticket open for this).

**Comment by kdschlosser** [comment]
OK so I got it working. YAY!!

Had to remove some of the Windows API calls as it was making things a lot more complicated then they really were. well actually the Windows API calls are still being made it is just happening in the C code for Python and not through ctypes which uses ffi.

I can remove most of the Windows API calls and substitute them for python calls. things like reading files and writing files. I already changed the MirrorFindFiles function to use Python to get the list of files and folders and I am able to get the handles to those from Python as well. I pass the handle GetFileInformationByHandle which populates the BY_HANDLE_FILE_INFORMATION structure and copy the information out of that structure into the WIN32_FIND_DATAW structure which gets passed to the FillFindData callback. 

The problem I was having with the desktop.ini file was due to how windows was returning the attributes from GetFileAttributes. the returned data type is DWORD and INVALID_FILE_ATTRIBUTES has a value of -1. problem there is a DWORD is an unsigned integer which INVALID_FILE_ATTRIBUTES is a signed integer. so I had to change the value of INVALID_FILE_ATTRIBUTES to 0xFFFFFFFF and that fixed that issue. I couldn't do a simple type conversion from unsigned to signed because it would change the returned value which is not what I wanted to do. I could have done the conversion in an if statement but that ends up being repetitive code so changing the value of INVALID_FILE_ATTRIBUTES seems like it is the better things to do.
